### PR TITLE
New db - semantically compatible with polygit1

### PR DESCRIPTION
**Pull request by garlicnation**

### CODE REVIEW

File: app-managed.go [hunk 1]
```diff
@@ -2,28 +2,72 @@
...
+
+	"google.golang.org/appengine"
+	"google.golang.org/cloud/compute/metadata"
 )
 
-func createMux() *echo.Echo {
-	// we're in a container on a Google Compute Engine instance so are not sandboxed anymore ...
-	runtime.GOMAXPROCS(runtime.NumCPU())
+func readGithubAPIKey() string {
+	md, err := metadata.Get("project/attributes/github-key")
```
Comment (justinfagnani) [inline app-managed.go]: 😱this is better than exceptions? </go-newb>

File: app-managed.go [hunk 2]
```diff
@@ -2,28 +2,72 @@
...
+	}
+	if md == "" {
+		log.Panic("Unable to read github api key!")
+		panic(err)
+	}
+	return md
+}
+
+func getBoltDb() *bolt.DB {
+	db, err := bolt.Open("/tmp/polygit.db", 0600, nil)
```
Comment (justinfagnani) [inline app-managed.go]: Do you have a memcache PR to follow up? I don't think a per-machine, on-disk cache is effective as memcache for cross-machine, and in-memory for per-machine.

File: app-managed.go [hunk 3]
```diff
@@ -2,28 +2,72 @@
...
-	// note: we don't need to provide the middleware or static handlers
-	// for the appengine vm version - that's taken care of by the platform
+func (*managedHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if strings.Contains(r.URL.EscapedPath(), "/components/") {
+		proxyGitPathRequest(w, r)
+	}
+}
 
-	return e
+func init() {
```
Comment (justinfagnani) [inline app-managed.go]: what's this?

File: app-managed.go [hunk 4]
```diff
@@ -2,28 +2,72 @@
...
+	db, err := bolt.Open("/tmp/polygit.db", 0600, nil)
+	if err != nil {
+		log.Panic("Unable to create bolt db")
+		panic(err)
+	}
+	return db
+}
 
-	e := echo.New()
+type managedHandler struct{}
```
Comment (justinfagnani) [inline app-managed.go]: what's the point of an empty struct?

File: app-standalone.go [hunk 5]
```diff
@@ -2,24 +2,81 @@
...
-	"github.com/labstack/echo"
-	"github.com/labstack/echo/middleware"
+	"expvar"
+	_ "net/http/pprof"
+	"time"
 )
 
-func createMux() *echo.Echo {
-	e := echo.New()
+import (
```
Comment (justinfagnani) [inline app-standalone.go]: what's the purpose of two separate import statements?

File: app-managed.go (hunk 1 above)
Comment (garlicnation) [inline app-managed.go]: Still not sure how I feel about it - this is the most lukewarm part of the language to me. There's also a panic/recover mechanism which works out almost like exceptions, with a higher runtime cost.

File: app-managed.go (hunk 2 above)
Comment (garlicnation) [inline app-managed.go]: This is effectively in-memory for serving, with the bonus of being able to survive the death/revival of the binary.

I have two ideas for how to scale this up. Ideally, we can leverage https://cloud.google.com/compute/docs/load-balancing/http/cdn and by setting the proper cache-control headers, get the platform to do the caching for us.

If that approach proves non-viable, the alternative I had in mind was memcache.

File: app-managed.go (hunk 3 above)
Comment (garlicnation) [inline app-managed.go]: Noise.

File: app-managed.go (hunk 4 above)
Comment (garlicnation) [inline app-managed.go]: It makes a convenient target to hang methods off of.

Ideas stolen from: http://dave.cheney.net/2014/03/25/the-empty-struct

File: app-standalone.go (hunk 5 above)
Comment (garlicnation) [inline app-standalone.go]: None. Cleaned up.

File: app-standalone.go [hunk 6]
```diff
@@ -2,24 +2,81 @@
...
+	if err != nil {
+		log.Panic("Unable to read github api key!")
+		panic(err)
+	}
+	log.Println("Got github api key!")
+	return string(apiKey)
+}
 
-	return e
+func getBoltDb() *bolt.DB {
```
Comment (justinfagnani) [inline app-standalone.go]: this appears to be the same as in app-managed...

File: app.go [hunk 7]
```diff
@@ -1,4 +1,32 @@
...
+
+// reference our echo instance and create it early
+// var e = createMux()
+
+// Check for an api key.
+var githubAPIKey string
+var getter repo.Getter
+
+func createGithubGetter() repo.Getter {
+	return getter
```
Comment (justinfagnani) [inline app.go]: what;s happening here? I don't quite understand how getter is set, or if it's a keyword, or what.

File: app.yaml [hunk 8]
```diff
@@ -1,26 +1,16 @@
-application: polygit            # defined when you create your app using google dev console
 module: default                 # see https://cloud.google.com/appengine/docs/go/
-version: 1                      # you can run multiple versions of an app and A/B test
+# version: 1                      # you can run multiple versions of an app and A/B test
```
Comment (justinfagnani) [inline app.yaml]: why is this commented out?

File: gitproxy.go [hunk 9]
```diff
@@ -1,50 +1,158 @@
...
+type getterFactory func() repo.Getter
+
+var whitelist = map[string]getterFactory{
+	"github.com": createGithubGetter,
 }
 
-func fetchFromGithub(user string, repo string, version string) (Repo, error) {
-	var r Repo
-	return r, nil
+func addAllValues(target []*repo.Repo, dict *map[string]*repo.Repo, resolved *map[string]*repo.VersionedRepo) []*repo.Repo {
```
Comment (justinfagnani) [inline gitproxy.go]: line length? what's the standard for Go?

Comment (justinfagnani) [inline gitproxy.go]: comment?

File: app.go (hunk 7 above)
Comment (garlicnation) [inline app.go]: The idea is that getter is set in -managed/-standalone.go

It's a package-level variable for the main package. This is a rough pass at a pattern to allow injection of request-specific context.

I can drop this indirection and just use getter directly in gitproxy.go:110, but this feels cleaner.

**Comment by garlicnation** [comment]
@justinfagnani PTAL

File: app.go (hunk 7 above)
Comment (justinfagnani) [inline app.go]: As a non-Go person, I just didn't see how this worked. No change needed.

File: app.yaml [hunk 10]
```diff
@@ -1,26 +1,16 @@
...
 api_version: go1                # used when appengine supports different go versions
 vm: true
-manual_scaling:
-  instances: 1
+automatic_scaling:
+  min_num_instances: 2
+  max_num_instances: 20
+  cool_down_period_sec: 60
+  cpu_utilization:
+    target_utilization: 0.5
```
Comment (justinfagnani) [inline app.yaml]: seems a little low?

File: vendor/vendor.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
+{
+	"comment": "",
+	"ignore": "test",
+	"package": [
+		{
+			"path": "github.com/Masterminds/semver",
+			"revision": "c4f7ef0702f269161a60489ccbbc9f1241ad1265",
```
Comment (justinfagnani) [inline vendor/vendor.json]: do these SHAs correspond to releases? can you select a tag instead?

File: vendor/github.com/Masterminds/semver/LICENSE.txt [hunk 12]
```diff
@@ -0,0 +1,20 @@
+The Masterminds
```
Comment (justinfagnani) [inline vendor/github.com/Masterminds/semver/LICENSE.txt]: Should you .gitignore vendor/ ?

File: urlconfig/parse.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	if configured.Name == "" {
+		candidate := repo.CompleteDependencyRepo(name, "*")
+		configured = *candidate
+	}
+	return configured
+}
+
+// Parse returns a config from a url
+func Parse(url string) (*ParsedRequest, error) {
```
Comment (justinfagnani) [inline urlconfig/parse.go]: to be clearer, I'd name the param `path`

File: urlconfig/parse.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+	// url = url[8:]
+	// Before this is config, after is a file.
+	splitURL := strings.SplitN(url, "/components/", 1)
```
Comment (justinfagnani) [inline urlconfig/parse.go]: and this, `splitPath`

File: urlconfig/parse.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+	if len(splitURL) != 2 {
+		return nil, errors.New("No /components/ in: " + url)
+	}
+
+	configMap := make(map[string]*repo.Repo)
+	// Figure out config
+	configs := strings.Split(splitURL[0], "/")
+	for idx := range configs {
+		c := configs[idx]
```
Comment (justinfagnani) [inline urlconfig/parse.go]: can we go with long names, like `config`?

File: urlconfig/parse.go [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+		var org, ver string
+		if len(splitC) == 2 {
+			ver = splitC[1]
+		} else if len(splitC) == 3 {
+			org = splitC[1]
+			ver = splitC[2]
+		} else {
+			return nil, errors.New("Unable to parse pattern: " + c)
+		}
+		if len(ver) > 1 {
```
Comment (justinfagnani) [inline urlconfig/parse.go]: https://golang.org/pkg/strings/#TrimPrefix

File: resolver/resolver.go [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+		newBytes := &bytes.Buffer{}
+		matchIndicies := r.FindSubmatchIndex(b)
+		newBytes.Write(b[matchIndicies[2]:matchIndicies[3]])
+		newBytes.WriteString(versionQuery)
+		newBytes.Write(b[matchIndicies[4]:matchIndicies[5]])
+		return newBytes.Bytes()
+	}
+}
+
+// Resolve resolves the contents of byte by replacing links to sibling resources
```
Comment (justinfagnani) [inline resolver/resolver.go]: remove first "resolve"?

Comment (justinfagnani) [inline resolver/resolver.go]: `byte` -> `content`?

File: resolver/resolver.go [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+// with resolved links to specific versions of dependencies.
+// dependencies is a map of the components dependencies.
+// versions is a map of the current pinned package versions
+func Resolve(content []byte, path string, dependencies map[string]*repo.Repo, versions map[string]string) []byte {
```
Comment (justinfagnani) [inline resolver/resolver.go]: Why do you need to rewrite imports? Don't all the imports in a page share the same base?

Comment (justinfagnani) [inline resolver/resolver.go]: I see, the call to this is commented out. Remove?

File: repo/repo.go [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+// Version represents a paired version and repo name
+type Version struct {
+	Repo    string
+	Version string
+}
+
+// Repo represents a git repository
+type Repo struct {
+	User    string `json:"user"`
```
Comment (justinfagnani) [inline repo/repo.go]: what's this syntax here?

Comment (garlicnation) [inline repo/repo.go]: https://eager.io/blog/go-and-json/

File: repo/repo.go [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Domain  string `json:"domain"`
+}
+
+// VersionedRepo contains a reference to a single version of a repository with optional
+// pinned dependency versions.
+type VersionedRepo struct {
+	Repo
```
Comment (justinfagnani) [inline repo/repo.go]: why does Repo not have a type? Is it implicit if there's a type name that matches the field name?

File: repo/repo.go (hunk 19 above)
Comment (garlicnation) [inline repo/repo.go]: The language specifies struct tags here: https://golang.org/ref/spec#Struct_types

File: vendor/vendor.json (hunk 11 above)
Comment (garlicnation) [inline vendor/vendor.json]: By convention `master` is a release branch and should never have breaking changes, this just records where on master we are.

File: repo/repo.go (hunk 19 above)
Comment (garlicnation) [inline repo/repo.go]: The key/value convention is semi-specified here: https://golang.org/pkg/reflect/#StructTag

It's considered public API and therefore stable.

File: repo/repo.go (hunk 20 above)
Comment (garlicnation) [inline repo/repo.go]: Repo isn't a conventional field, it's being embedded into VersionedRepo.

Docs: https://golang.org/doc/effective_go.html#embedding

example:
```
x := VersionedRepo{}
// Valid
x.Name = "boo"
//Invalid
x.Repo = Repo{}
```

File: vendor/github.com/Masterminds/semver/LICENSE.txt (hunk 12 above)
Comment (garlicnation) [inline vendor/github.com/Masterminds/semver/LICENSE.txt]: vendor/ is used to simulate a monorepo and ease deploy/build, so I'm keep it intentionally.

**Comment by garlicnation** [comment]
@justinfagnani PTAL.

File: gitproxy.go [hunk 21]
```diff
@@ -1,50 +1,163 @@
...
+type getterFactory func() repo.Getter
+
+var whitelist = map[string]getterFactory{
+	"github.com": createGithubGetter,
 }
 
-func fetchFromGithub(user string, repo string, version string) (Repo, error) {
-	var r Repo
-	return r, nil
+// addAllValues adds all repos from  `dict` into `target` that aren't already in `resolved`
```
Comment (justinfagnani) [inline gitproxy.go]: `dict` -> `dependencies`?

File: gitproxy.go [hunk 22]
```diff
@@ -1,50 +1,163 @@
...
+			Name:    v.Name,
+			Version: v.Version,
+			Domain:  v.Domain}
+		target = append(target, r)
+	}
+	return target
 }
 
-type repoFetcher func(string, string, string) (Repo, error)
+func redirectIfCaps(w http.ResponseWriter, r *http.Request) bool {
```
Comment (justinfagnani) [inline gitproxy.go]: coment?

File: gitproxy.go [hunk 23]
```diff
@@ -1,50 +1,163 @@
...
+	loadedRepo, err := getter.GetRepo(parsedReq.Repo.User, parsedReq.Repo.Name, parsedReq.Repo.Version)
+	log.Printf("GetRepo took %s", time.Since(start))
+	if err != nil || r == nil {
+		sendError(w, err)
+		return
+	}
+	// Get dependency information transitively.
+	depQueue := []*repo.Repo{}
+	depQueue = addAllValues(depQueue, loadedRepo.Dependencies, nil)
+	// for len(depQueue) > 0 {
```
Comment (justinfagnani) [inline gitproxy.go]: remove commented out code

**Comment by justinfagnani** [comment]
Ok, I've done all I can here :) Thanks for answering my Go questions. LGTM

**Pull request merged**